ttl/ttlworker/job_manager.go

```diff
@@ -719,6 +573,11 @@ func (m *JobManager) lockNewJob(ctx context.Context, se session.Session, table *
 			return errors.Wrapf(err, "execute sql: %s", sql)
 		}
 
+		// if the job already exist, don't need to submit scan tasks
+		if jobExist {
+			return nil
+		}
+
 		ranges, err := table.SplitScanRanges(ctx, m.store, splitScanCount)
```

Comment:
I'm not sure whether there will be some problems for inserting job and tasks in one transaction. If one table is very big, `SplitScanRanges` may take much more time to execute and it increase the possibility of the conflict.

Maybe the better implement is to lock job in one transaction first, and then insert the tasks in other transactions separately.

Reply:
> it increase the possibility of the conflict.

IIUC, the problem is not about conflict. It's the resource waster because many tidb nodes will run the same logic, all of them will split the table but only one for them will success. If the `SplitScanRanges` is time consuming, I think it's OK to only run it when the node can make sure the task can be locked successfully.

On the other hand, moving it out of the transaction will make the state more complicated. The assumption "if there is a jobID, I could query all tasks with the id" will fail, because the tasks could still be empty. Though it seems fine now, it will bring some burden to the developers.

Could we try to use pessimistic transaction and `select .. for update nowait` to make the lock fail fast? I'll have a try.

Reply:
`select .. for update nowait` works pretty well :+1: